torchdata/datapipes/iter/util/decompressor.py

```diff
@@ -114,3 +121,106 @@ def __new__(
+    StreamWrapper and isn't intended to be used outside decompressor.py. It only supports the
+    specific operations expected by StreamWrapper.
+    """
+
+    def __init__(self, file) -> None:
+        self._decompressor = zlib.decompressobj()
+        self._file = file
+
+        # Stores decompressed bytes leftover from a call to __next__: since __next__ only returns
+        # up until the next newline, we need to store the bytes from after the newline for future
+        # calls to read or __next__.
+        self._buffer = bytearray()
+
+        # Whether or not self._file still has bytes left to read
+        self._file_exhausted = False
+
+    def read(self, size: int = -1) -> bytearray:
+        if size < 0:
+            return self.readall()
+
+        if not size:
+            return bytearray()
+
+        result = self._buffer[:size]
+        self._buffer = self._buffer[size:]
+        while len(result) < size and self._compressed_bytes_remain():
+            # If decompress was called previously, there might be some compressed bytes from a previous chunk
+            # that haven't been decompressed yet (because decompress() was passed a max_length value that
+            # didn't exhaust the compressed bytes in the chunk). We can retrieve these leftover bytes from
+            # unconsumed_tail:
+            chunk = self._decompressor.unconsumed_tail
+            # Let's read compressed bytes in chunks of io.DEFAULT_BUFFER_SIZE because this is what python's gzip
+            # library does:
+            # https://github.com/python/cpython/blob/a6326972253bf5282c5bf422f4a16d93ace77b57/Lib/gzip.py#L505
+            if len(chunk) < io.DEFAULT_BUFFER_SIZE:
+                compressed_bytes = self._file.read(io.DEFAULT_BUFFER_SIZE - len(chunk))
+                if compressed_bytes:
+                    chunk += compressed_bytes
+                else:
+                    self._file_exhausted = True
+            decompressed_chunk = self._decompressor.decompress(chunk, max_length=size - len(result))
+            result.extend(decompressed_chunk)
+        if not self._compressed_bytes_remain() and not self._decompressor.eof:
+            # There are no more compressed bytes available to decompress, but we haven't reached the
+            # zlib EOF, so something is wrong.
+            raise EOFError("Compressed file ended before the end-of-stream marker was reached")
+        return result
+
+    def readall(self):
+        """
+        This is just mimicking python's internal DecompressReader.readall:
+        https://github.com/python/cpython/blob/a6326972253bf5282c5bf422f4a16d93ace77b57/Lib/_compression.py#L113
+        """
+        chunks = []
+        # sys.maxsize means the max length of output buffer is unlimited,
+        # so that the whole input buffer can be decompressed within one
+        # .decompress() call.
+        data = self.read(sys.maxsize)
+        while data:
```

Comment:
I wasn't sure if python 3.7 support was desired, so I didn't use the walrus operator here.